src/System.CommandLine.StaticCompletions/CompletionsCommand.cs

```diff
@@ -0,0 +1,105 @@
+    { }
+
+    private CompletionsCommand(Dictionary<string, IShellProvider> shellMap)
+        : base("completions", "Commands for generating and registering completions for supported shells")
+    {
+        var shellArg = new CliArgument<IShellProvider>("shell")
+        {
+            Description = "The shell for which to generate or register completions",
+            Arity = ArgumentArity.ZeroOrOne,
+            // called when no token is presented at all
+            DefaultValueFactory = (argResult) => LookupShellFromEnvironment(shellMap),
+            // called for all other scenarios
+            CustomParser = (argResult) =>
+            {
+                return argResult.Tokens switch
+                {
+                // shouldn't be required because of the DefaultValueFactory above
+                [] => LookupShellFromEnvironment(shellMap),
+                [var shellToken] => shellMap[shellToken.Value],
+                    _ => throw new InvalidOperationException("Unexpected number of tokens")
+                };
+            }
+        };
+
+        shellArg.AcceptOnlyFromAmong(shellMap.Keys.ToArray());
+        Subcommands.Add(new GenerateScriptCommand(shellArg));
+        Subcommands.Add(new RegisterScriptCommand(shellArg));
+
+        static IShellProvider LookupShellFromEnvironment(Dictionary<string, IShellProvider> shellMap)
+        {
+            if (OperatingSystem.IsWindows())
+            {
+                return shellMap[PowershellShellProvider.PowerShell];
+            }
+            var shellPath = Environment.GetEnvironmentVariable("SHELL");
+            if (shellPath is null)
+            {
+                throw new InvalidOperationException("Could not determine the shell from the environment");
+            }
+            var shellName = Path.GetFileName(shellPath);
+            if (shellMap.TryGetValue(shellPath, out var shellProvider))
+            {
+                return shellProvider;
+            }
+            else
+            {
+                throw new InvalidOperationException($"Shell '{shellPath}' is not supported");
+            }
+
+            return shellProvider;
+        }
+    }
+}
+
+public class GenerateScriptCommand : CliCommand
+{
+    public GenerateScriptCommand(CliArgument<IShellProvider> shellArg)
+        : base("script", "Generate the completion script for a supported shell")
+    {
+        Arguments.Add(shellArg);
+        SetAction(args =>
+        {
+            var shell = args.GetValue(shellArg);
+            if (shell is null)
+            {
+                throw new InvalidOperationException("No shell was provided");
+            }
+            var script = shell.GenerateCompletions(args.RootCommandResult.Command);
+            args.Configuration.Output.Write(script);
+        });
+    }
+}
+
+public class RegisterScriptCommand : CliCommand
```

Comment:
Can you explain the value of this command? Seems like it does strictly less than GenerateScript

Reply:
This is some ground work that I should probably delete because it's vestigial at this time. The thinking is we should have an explicit command that users run to generate the completion script for their shell, and that makes total sense to do. Users that want completions can open their Powershell $PROFILE and add `dotnet completions generate pwsh | Out-String | Invoke-Expression` to it and have up-to-date completions each time they load their shell. This can of course be optimized, etc. but this is the main idea of many shell-completion frameworks that exist today.

However, you still need to get users to _do_ that step. This is why some other frameworks have an 'install' and 'uninstall' semantic. This looks something like `dotnet completions register pwsh`, and the logic would be, roughly:

* determine the shell the user is requesting for
* ask the shell provider to see if the dotnet completions are registered
  * this will be heuristic-based, likely, and require probing contents of files like $PROFILE for pwsh, .zshrc for Zsh, etc.
* if not, inject a section to call the `dotnet completions generate` command into the shell configuration files
* (optionally) support uninstalling as well - do the same but find and remove the special section

The 'fuzziness' of this command is why I haven't implemented it yet. But once it exists, we could then have a discussion about if the SDK first-run experience should set up this completion registration automatically - we know today that so very very few of our users use completions, and having them set up would be a huge usability increase - as well as make it so that any fixes/enhancement to the completions we might make would be automatically applied!

Reply:
I like the idea of a dotnet completions register pwsh command, but even more than that, I like the idea of making this part of the first-run welcome message but in a slightly different way. I'm envisioning two options:
1. A message like "If you want to see what options are available, run 'dotnet completions register pwsh'."
2. Optionally pause before executing their command with a question like "do you want completions? y/n" and let them choose.

I think something like that would dramatically increase usage, and I do agree with "huge usability increase" _if_ people start using it...but I also don't like the idea of opting people into something without telling them. I actually do see an argument in favor of not having completions, namely that it can take up a lot of screen real estate, and it may not ultimately be helpful if we do a poor job naming our subcommand/options, and people feel the need to go online to read more about them anyway. As we're busy telling people how to opt in (or giving them the y/n option), we can make sure to mention something like "if you ever don't want this anymore, run 'dotnet completions unregister pwsh'." Though I'll admit that unregister might go the way of unregister in MSBuildLocator in becoming vestigial, I think it's positive to offer it as an option initially. It makes it less scary to opt into something if I know I can change my mind later.

As a small aside, I like install/uninstall more than register/unregister, but maybe that's just me.